### ISSUE TITLE
JInput::getFloat() does not interpret ".x" values correctly

### ISSUE THREAD

**Issue by JTBlum** [joomla/joomla-cms#14156]
### Steps to reproduce the issue
$in = JApplicationCms::getInstance()->input;
die($in->getFloat('testFloat',0.0));

Navigate to the page and add the $_GET parameter testFloat=.25

### Expected result
Blank page except for the text ".25"

### Actual result
Blank page except for the text "25"

### System information (as much as possible)
Joomla! Version	Joomla! 3.6.5 Stable [ Noether ] 1-December-2016 22:46 GMT
Web Server	Apache/2.4.6 (CentOS) OpenSSL/1.0.1e-fips mod_fcgid/2.3.9 PHP/7.0.16
PHP Built On	Linux jbdev01 3.10.0-327.22.2.el7.x86_64 #1 SMP Thu Jun 23 17:05:11 UTC 2016 x86_64
PHP Version	7.0.16

### Additional comments
Starting the value with a zero (0.25) returns the expected result

**Comment by JTBlum** [comment]
This also happens with JRequest::getFloat<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14156">issues.joomla.org/tracker/joomla-cms/14156</a>.</sub>

**Comment by C-Lodder** [comment]
Current regex for `float` is:
```
$pattern = '/[-+]?[0-9]+(\.[0-9]+)?([eE][-+]?[0-9]+)?/';
```

which as you may be able to see, doesn't allow the value to start with a decimal. It would have to be changed to:
```
$pattern = '/[-+.]?[0-9]+(\.[0-9]+)?([eE][-+]?[0-9]+)?/';
                ^
```

However looking at the [PHP Documentation](http://php.net/manual/en/language.types.float.php), I think this is expected.

Perhaps the `INT/INTEGER` filter should be updated to allow a starting decimal place.

@mbabker ?

**Comment by JTBlum** [comment]
I would think the actual change would be as follows:

<pre>$pattern = '/[-+]?[0-9]*(\.[0-9]+)?([eE][-+]?[0-9]+)?/';
                       ^</pre><hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14156">issues.joomla.org/tracker/joomla-cms/14156</a>.</sub>

**Comment by mbabker** [comment]
My opinion.  If PHP can't natively convert .25 to 0.25 I would be careful with adding support for that to our filtering library.  As it's supposed to be a filter, I wouldn't add too much magic to make it work as more.

That said, if it does handle it fine, feel free to submit a PR with appropriate unit test coverage (just adding another case to the existing array of cases).  As security sensitive code, I push a little more for changes in JFilterInput to be backed with unit tests than I do for other classes.

**Comment by JTBlum** [comment]
PHP converts .25 to 0.25 just fine, and this is supported by the documentation cited by C-Lodder.  Check out the DNUM pattern, quoted from that page:
<pre>LNUM          [0-9]+
DNUM          ([0-9]*[\.]{LNUM}) | ({LNUM}[\.][0-9]*)
EXPONENT_DNUM [+-]?(({LNUM} | {DNUM}) [eE][+-]? {LNUM})</pre>

If you really need a code example, see:

echo floatval('.25') === 0.25 ? 'Good' : 'Bad';
echo "\n&lt;br/&gt;\n";
echo floatval('.25') === .25 ? 'Good' : 'Bad';

This produces the following string on all versions I tested:
Good 
Good<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14156">issues.joomla.org/tracker/joomla-cms/14156</a>.</sub>

**Comment by mbabker** [comment]
Then someone feel free to submit the pull request.  I just wanted to make sure we aren't adding more magic behavior is all.

**Comment by tonypartridge** [comment]
@JTBlum have you had a chance to do a pull request?

**Comment by JTBlum** [comment]
I’m afraid I’m not experienced enough with Github or with the Joomla core itself to make the change and be at all confident I haven’t broken something else.  I was hoping someone else would address the issue.

**Comment by JTBlum** [comment]
@Franz: Thanks for confirming this.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14156">issues.joomla.org/tracker/joomla-cms/14156</a>.</sub>

**Comment by carcam** [comment]
@JTBlum I have submitted the PR with the code you propose in https://github.com/joomla/joomla-cms/issues/14156#issuecomment-281360112 as after testing the ReGex it seems right.

**Comment by franz-wohlkoenig** [comment]
closed as having PR #15736<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14156">issues.joomla.org/tracker/joomla-cms/14156</a>.</sub>
